### PR TITLE
Fix pointer arithmetic in EKF logging

### PR DESCRIPTION
**Pull request by PeterDuerr**
* sizeof returns byte, but buf.estimator_status.states and
  buf.estimator_status.covariances are float pointers
* Code is not tested

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Have you actually seen failure as a result of this? On my first pass the logic seems right - its pointer arithmetic on float pointers resulting in correct float offsets.

Your comment about sizeof does not apply here. Did you maybe intent to fix the sizeof logic a few lines above?

File: src/modules/sdlog2/sdlog2.c [hunk 1]
```diff
@@ -1793,7 +1793,7 @@ int sdlog2_thread_main(int argc, char *argv[])
 			log_msg.msg_type = LOG_EST1_MSG;
 			unsigned maxcopy1 = ((sizeof(buf.estimator_status.states) - maxcopy0) < sizeof(log_msg.body.log_EST1.s)) ? (sizeof(buf.estimator_status.states) - maxcopy0) : sizeof(log_msg.body.log_EST1.s);
```
Comment (LorenzMeier) [inline src/modules/sdlog2/sdlog2.c]: did you mean to change this line to:
```C
unsigned maxcopy1 = ((sizeof(*buf.estimator_status.states) - maxcopy0)
```

File: src/modules/sdlog2/sdlog2.c [hunk 2]
```diff
@@ -1805,7 +1805,7 @@ int sdlog2_thread_main(int argc, char *argv[])
 			log_msg.msg_type = LOG_EST3_MSG;
 			unsigned maxcopy3 = ((sizeof(buf.estimator_status.covariances) - maxcopy2) < sizeof(log_msg.body.log_EST3.cov)) ? (sizeof(buf.estimator_status.covariances) - maxcopy2) : sizeof(log_msg.body.log_EST3.cov);
```
Comment (LorenzMeier) [inline src/modules/sdlog2/sdlog2.c]: Did you mean to change this to:
```C
unsigned maxcopy3 = ((sizeof(*buf.estimator_status.covariances) - maxcopy2)
```

**Comment by LorenzMeier** [comment]
It seems the right fix would be to fix the maxcopy size estimates, not to change the copy command itself.

**Comment by PeterDuerr** [comment]
> Have you actually seen failure as a result of this? On my first pass the logic seems > right - its pointer arithmetic on float pointers resulting in correct float offsets.

Yes, the current code does not work correctly (see explanation below).

> Your comment about sizeof does not apply here. Did you maybe intent to fix the 
> sizeof logic a few lines above?

Sorry, I should have explained better. The maxcopy size estimates compute the number of *bytes* that should be copied (the third argument of memset and memcpy should be a number of bytes).  The second argument to memcpy (the source pointer) is then computed as the sum of a float pointer and an unsigned number of bytes, which leads to the wrong address being computed. I cast to a char pointer (and not to a void*) in order to warn a future reviewer who might be tempted to remove the typecast, but maybe the code should be refactored more aggressively anyway. 

> Did you mean to change this line to:
> unsigned maxcopy1 = ((sizeof(*buf.estimator_status.states) - maxcopy0)

> Did you mean to change this to:
> unsigned maxcopy3 = ((sizeof(*buf.estimator_status.covariances) - maxcopy2)

No, the original code looks correct to me.

> It seems the right fix would be to fix the maxcopy size estimates, not to change the copy command itself.

One could of course compute maxcopy as a float offset, but that would require multiplying by sizeof(float) for the third argument to memset / memcpy. Alternatively one could divide by sizeof(float) in the computation of the source pointer, but I (perhaps mistakenly) considered casting to char more readable.

**Comment by LorenzMeier** [comment]
Ok, I see it now, thanks!

**Comment by LorenzMeier** [comment]
Cherry-picked and applied, thanks!

**Pull request closed without merging**